### ISSUE TITLE
Incorrect Gradient for construction of Matrix

### ISSUE THREAD

**Issue by half-potato** [shader-slang/slang#3524]
Here is a unit test that will produce an incorrect gradient:
```
import jax
import jax.numpy as jnp
import numpy as np
import slangpy
from pathlib import Path

def jquatToMat3(q):
    r = q[0]
    x = q[1]
    y = q[2]
    z = q[3]
    return jnp.array([
                        1.0 - 2.0 * (y * y + z * z),
                        2.0 * (x * y - r * z),
                        2.0 * (x * z + r * y),
                        2.0 * (x * y + r * z),
                        1.0 - 2.0 * (x * x + z * z),
                        2.0 * (y * z - r * x),
                        2.0 * (x * z - r * y),
                        2.0 * (y * z + r * x),
                        1.0 - 2.0 * (x * x + y * y),

    ]).reshape(3, 3).T

fn = slangpy.loadModule("test.slang")
quat = np.array([0.9811850346541486, 0.13331543321474051, 0.07252929475418773, 0.1193416291155902])
rayd = np.ones((3))
scale = np.ones((3))
def test(quat, scale, vec):
    Trayd = jquatToMat3(quat).T @ jnp.diag(1/scale) @ vec
    # Trayd = jquatToMat3(quat).T @  vec
    return (Trayd * Trayd).sum()

def test3(quat, scale, vec):
    Trayd = jquatToMat3(quat).T @ jnp.diag(1/scale) @ vec
    # Trayd = jquatToMat3(quat).T @  vec
    return Trayd

print("\n\nUnit test 3")
vec, dtest3 = jax.vjp(test3, quat, scale, rayd.reshape(-1))
print(vec, dtest3(np.ones((3))))
args = (quat.tolist(), scale.tolist(), rayd.reshape(-1).tolist())
print(fn.test3(*args), fn.dtest3(*args, [1.0, 1.0, 1.0]))

print("\n\nUnit test 1")
vec, dtest = jax.vjp(test, quat, scale, rayd.reshape(-1))
print(vec, dtest(1.0))
print(fn.test(*args), fn.dtest(*args, 1.0))

```

test.slang:
```
[Differentiable]
float3x3 quat2mat(float4 quat)
{
    let r = quat[0];
    let x = quat[1];
    let y = quat[2];
    let z = quat[3];
    float3x3 R = {
        1.0 - 2.0 * (y * y + z * z),
        2.0 * (x * y - r * z),
        2.0 * (x * z + r * y),

        2.0 * (x * y + r * z),
        1.0 - 2.0 * (x * x + z * z),
        2.0 * (y * z - r * x),

        2.0 * (x * z - r * y),
        2.0 * (y * z + r * x),
        1.0 - 2.0 * (x * x + y * y)
    };
    return R;
}

[Differentiable]
float3x3 diag(float3 vec)
{
    return {
      vec.x, 0, 0,
      0, vec.y, 0,
      0, 0, vec.z,
    };
}


[TorchEntryPoint]
[Differentiable]
float test(float4 quat, float3 scale, float3 vec)
{
    /*
    float3x3 invS = diag({
        1/max(scale.x, 1e-8),
        1/max(scale.y, 1e-8),
        1/max(scale.z, 1e-8),
    });
    */
    // float3x3 invS = float3x3(1.0);
    float3x3 invS = diag({1.f, 1.f, 1.f});
    float3x3 mat = mul(transpose(quat2mat(quat)), invS);
    //float3x3 mat = transpose(quat2mat(quat));
    let Trayd = mul(vec, mat);
    return dot(Trayd, Trayd);
}

[TorchEntryPoint]
[Differentiable]
float3 test3(float4 quat, float3 scale, float3 vec)
{
    float3x3 invS = {
        {1/scale.x, 0.f, 0.f},
        {0.f, 1/scale.x, 0.f},
        {0.f, 0.f, 1/scale.z},
    };
    float3x3 mat = mul(transpose(quat2mat(quat)), invS);
    //float3x3 mat = transpose(quat2mat(quat));
    let Trayd = mul(vec, mat);
    return Trayd;
}

struct TestOut{
    float4 quat;
    float3 scale;
    float3 vec;
};

[TorchEntryPoint]
TestOut dtest(float4 quat, float3 scale, float3 vec, float diff)
{
    TestOut out;
    var dquat = diffPair(quat, {});
    var dscale = diffPair(scale, {});
    var dvec = diffPair(vec, {});
    bwd_diff(test)(dquat, dscale, dvec, diff);
    out.quat = dquat.d;
    out.scale = dscale.d;
    out.vec = dvec.d;
    return out;
}

[TorchEntryPoint]
TestOut dtest3(float4 quat, float3 scale, float3 vec, float3 diff)
{
    TestOut out;
    var dquat = diffPair(quat, {});
    var dscale = diffPair(scale, {});
    var dvec = diffPair(vec, {});
    bwd_diff(test3)(dquat, dscale, dvec, diff);
    out.quat = dquat.d;
    out.scale = dscale.d;
    out.vec = dvec.d;
    return out;
}
```

Here is the output

```
Unit test 3
CUDA backend failed to initialize: Found CUDA version 12010, but JAX was built against version 12020, which is newer. The copy of CUDA that is installed must be at least as new as the version against which JAX was built. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[0.9202897 0.9451974 1.1223496] (Array([ 0.        , -0.29903978,  0.43039393, -0.13135403], dtype=float32), Array([-1.104016  , -1.000041  , -0.88377976], dtype=float32), Array([1.104016  , 1.000041  , 0.88377976], dtype=float32))
(0.9202896952629089, 0.9451974034309387, 1.1223496198654175) ((0.0, -0.299039751291275, 0.43039390444755554, -0.13135409355163574), (-1.8654870986938477, 0.0, -1.1223496198654175), (1.1040159463882446, 1.0000410079956055, 0.883779764175415))


Unit test 1
2.9999998 (Array([0.04773754, 0.00648645, 0.00352854, 0.00580615], dtype=float32), Array([-1.9999998, -1.9999999, -2.       ], dtype=float32), Array([1.9999998, 1.9999999, 2.       ], dtype=float32))
3.0 ((0.04773754999041557, 0.006486356258392334, 0.0035283565521240234, 0.005806326866149902), (0.0, 0.0, 0.0), (1.999999761581421, 1.9999998807907104, 2.0))
```
Here is the derivative wrt the scales parameter from jax:
`Array([-1.104016  , -1.000041  , -0.88377976], dtype=float32)`
Here is slang:
`(-1.8654870986938477, 0.0, -1.1223496198654175)`

**Comment by saipraveenb25** [comment]
Hi @half-potato, thank you for raising this issue!

I took a closer look, and I don't think there's anything wrong with the gradient generation.

I believe the JAX code and the Slang code are not computing the same thing. 

There are two differences:

1. There is a typo in the Slang code in the `test3()` method where the diagonal matrix has `scale.x` repeated twice instead of using `scale.y` for the second row. 
2. The `test3()` method in the python code should use `Trayd = vec.T @ (jquatToMat3(quat) @ jnp.diag(1/scale))` to match the sequence of multiplications in the Slang code rather than `Trayd = jquatToMat3(quat).T @ jnp.diag(1/scale) @ vec` which is a different sequence of operations. It is possible that this error occured because when `scale` is all 1s, the two operations produce the same primal output (but have different gradients)

After making the above corrections, both the JAX and Slang derivatives were the same.

Please let me know if this resolves your issue.

**Comment by half-potato** [comment]
Damn, using all ones for the scale has screwed me up twice. Sorry for wasting time with this bug. Thanks for taking a look.
